security/security-api/custom-identity-store/README.md

```diff
@@ -0,0 +1,175 @@
+            // the Identity Store
+            // Password internally gets verified using the PasswordHash specified while defining DatabaseIdentityStore.
+            CredentialValidationResult result = identityStoreHandler.validate(
+                new UsernamePasswordCredential(name, password));
+
+            if (result.getStatus() == VALID) {
+                // Communicate the details of the authenticated user to the
+                // container. In many cases the underlying handler will just store the details
+                // and the container will actually handle the login after we return from
+                // this method.
+                return httpMessageContext.notifyContainerAboutLogin(
+                    result.getCallerPrincipal(), result.getCallerGroups());
+            } else {
+                return httpMessageContext.responseUnauthorized();
+            }
+        }
+
+        return httpMessageContext.doNothing();
+    }
+
+}
+```
+Following snippet from `TestAuthenticationMechanism` delegates the validation task to underlying `DefaultIdentityStoreHandler`, which in turn validates the credentials against `TestIdentityStore` as defined in this application.
+
+```java
+CredentialValidationResult result = identityStoreHandler.validate(
+    new UsernamePasswordCredential(name, password));
+```
+
+On successful authorization, container gets notified with caller details.
+
+```java
+return httpMessageContext.notifyContainerAboutLogin(
+    result.getCallerPrincipal(), result.getCallerGroups());
+```
+If credentials are not valid, an unauthorized response is sent back to the caller with status `401` indicating that authentication has failed with provided credentails.
+```java
+return httpMessageContext.responseUnauthorized();
+```
+
+When a request is made to the application, the roles the user is in, gets returned as part of the repsonse.
+
+```java
+@DeclareRoles({ "foo", "bar", "kaz" })
+@WebServlet("/servlet")
+public class Servlet extends HttpServlet {
+
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+
+        String webName = null;
+        if (request.getUserPrincipal() != null) {
+            webName = request.getUserPrincipal().getName();
+        }
+
+        response.getWriter().write("web username: " + webName + "\n");
+
+        response.getWriter().write("web user has role \"foo\": " + request.isUserInRole("foo") + "\n");
+        response.getWriter().write("web user has role \"bar\": " + request.isUserInRole("bar") + "\n");
+        response.getWriter().write("web user has role \"kaz\": " + request.isUserInRole("kaz") + "\n");
+    }
+
+}
+```
+Notice that, the container itself needs to be made aware of the supported roles and it is done with the help of `@DeclareRoles` annotation as shown above.
```

Comment:
Note that the container needs to be made aware of the supported roles, which is achieved with the help of the '@DeclareRoles' annotation as shown above.